specs/feature-flags.md

```diff
@@ -1,248 +1,448 @@
+   * @added 2025-01-20
+   * @owner @username
+   * @removal 2025-07-20
+   */
+  BATCH_OPERATIONS: z.boolean(),
+
+  /**
+   * Experimental tool with configurable limits
+   * @added 2025-02-01
+   * @owner @username
+   * @removal 2025-05-01
+   */
+  EXPERIMENTAL_TOOL: z.object({
+    enabled: z.boolean(),
+    max_items: z.number().min(1).max(1000),
+  }),
+} as const;
+```
 
-**[NEEDS CLARIFICATION: Where is this metadata stored? In code? In KV? In a separate registry file?]**
+**Benefits of JSDoc approach:**
+- Structured and parsable (enables future tooling)
+- Co-located with schemas (single source of truth)
+- Familiar format for TypeScript developers
+- Potential to generate documentation automatically
 
 ### Flag Naming Conventions
 
-**[NEEDS CLARIFICATION]** Flag names should follow a consistent convention:
-- Use SCREAMING_SNAKE_CASE
-- Prefix with feature area (e.g., `SEARCH_ADVANCED_RANKING`, `STORAGE_BATCH_OPERATIONS`)
-- Avoid negative names (use `FEATURE_ENABLED` not `FEATURE_DISABLED`)
+Flag names MUST follow these conventions:
+- **Format:** SCREAMING_SNAKE_CASE
+- **Prefix:** Feature area when helpful (e.g., `SEARCH_ADVANCED_RANKING`, `STORAGE_BATCH_DELETE`)
+- **Positive names:** Use `FEATURE_ENABLED` not `FEATURE_DISABLED` (avoid negation)
+- **Length:** Keep under 40 characters for readability
+
+---
 
-**[NEEDS CLARIFICATION: Should naming be enforced programmatically? Should there be a maximum flag name length?]**
+## Flag Management
 
-## Flag Usage
+### Creating Flags
 
-**[NEEDS CLARIFICATION: How should code access and check flag values?]**
+Process for adding a new flag:
 
-### Accessing Flags
+1. **Define schema** in the flag schema registry (TypeScript file)
+2. **Add JSDoc metadata** (@description, @added, @owner, @removal)
+3. **Set default value** in the `production` flag set in KV
+4. **Override in development** flag set in KV if needed for testing
+5. **Use flag in code** with proper type checking
+6. **Deploy code** that checks the flag
+7. **Verify flag works** in development environment
 
-**[NEEDS CLARIFICATION]** Code must access flags from request context since KV bindings are only available in env:
+**Schema validation requirement:**
+All flag set updates (manual or automated) MUST validate against the schema registry. Invalid flag values MUST be rejected.
 
-```typescript
-// Pseudocode example - NEEDS CLARIFICATION
-async function handleRequest(request: Request, env: Env): Promise<Response> {
-  const flagSet = await determineFlagSet(request, env);
-  const flags = await loadFlags(flagSet, env);
+**[DEFERRED]** CLI tooling for flag creation is a future enhancement. Initial implementation uses manual KV updates (via wrangler CLI or dashboard).
 
-  if (flags.ADVANCED_SEARCH) {
-    // Feature-flagged code path
-  } else {
-    // Default code path
+### Updating Flags
+
+Process for changing flag values at runtime:
+
+1. **Update flag set** in KV (modify JSON blob for target flag set)
```

Comment:
While we don't need to mandate specifics we need to imply more of an automated process. Ideally there would be a dashboard with a UI although maybe for simplicity purposes it's just some scripts that hit the cloudflare API or even an endpoint of the application, this way schema validation can be enforced.